### PR TITLE
CI text style check

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature:** This PR introduces a new script that checks the formatting of text files, as requested in the discord server by reviewers.

## Feature Details
This script is integrated into Github Workflows to check any changes made by pull requests. Currently, it checks the format of all `.txt` files in the `data` directory.

I made this into a draft PR so we can discuss what needs to be checked here, and if the current approach is correct. As I'm not somebody who has to deal with the syntax of missions / game content often, I don't really know what other common issues there are, or if they can be checked in such a minimalist way.

This PR is not meant to implement a full data parser for the game to check mission syntax.

## Tasks
- [x] Check for copyright headers
- [x] Check for non-ASCII characters
- [x] Check for multiple spaces (except before comments, because that is used extensively for some reason)
- [x] Check indentation on empty lines (follows Sam's suggesiton with non-indented empty lines)
- [x] Check that not more than a single level of indentation is added between two non-empty lines
- [x] Check that only tabs are used as indentation
- [x] Check for unnecessary space before closing backtick (might need some more work)
- [ ] Add even more checks based on requests

All of the existing checks have found issues in the data files, which I'm keeping in a draft PR for now.

## Testing Done
Some testing was done to ensure that the script is functional.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -515,3 +515,22 @@ jobs:
...
     - run: editorconfig-checker
+
+
+  check-content-style:
+    needs: changed
+    if: needs.changed.outputs.data == 'true'
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
```
Comment (quyykk) [inline .github/workflows/ci.yml:526]:
```suggestion
      uses: actions/checkout@v3
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -515,3 +515,22 @@ jobs:
...
+    - name: set up python environment
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.10.6'
+    - name: install python packages
+      run: |
+        python -m pip install --upgrade pip
```
Comment (quyykk) [inline .github/workflows/ci.yml:533]: Do we need this?

**Review by quyykk — COMMENTED** [review]
I don't have any other comment. So I think this looks good

File: .github/workflows/ci.yml (hunk 2 above)
Comment (tibetiroka) [reply .github/workflows/ci.yml:533]: No idea.

File: utils/check_content_style.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+# Checks the format of all txt files.
+def check_content_style():
+	files = glob.glob('data/**/*.txt', recursive=True)
+	files.sort()
+	for file in files:
+
+		f = open(file, "r")
+		lines = f.readlines()
+		lines = [line.removesuffix('\n') for line in lines]
```
Comment (tehhowch) [inline utils/check_content_style.py:46]: Why remove the line ending?

This script currently uses the "universal newlines" mode of `open`, so CRLF will get translated to LF even if the host platform is not Windows. That seems like a scenario we would want to catch, and may be catchable by setting `newline=''` in the call to `open` and then comparing against `os.linesep`

Another scenario we'd probably want to catch would be the inclusion of the UTF BOM (byte-order mark), but you'd likely need to open the files in binary mode for that.

File: utils/check_content_style.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+		check_copyright(file, lines)
+		check_ascii(file, lines)
+		check_regexes(file, lines)
+		check_indents(file, lines)
```
Comment (tehhowch) [inline utils/check_content_style.py:51]: You shouldn't need to pass the filename into these functions; you can compose the errors / annotations from this calling context.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -515,3 +515,20 @@ jobs:
...
+    needs: changed
+    if: needs.changed.outputs.data == 'true'
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
+    - name: set up python environment
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.10.6'
```
Comment (tehhowch) [inline .github/workflows/ci.yml:530]:
```suggestion
        python-version: '3.x'
```

File: utils/check_content_style.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+
+# Checks the format of all txt files.
+def check_content_style():
+	files = glob.glob('data/**/*.txt', recursive=True)
+	files.sort()
+	for file in files:
+
+		f = open(file, "r")
+		lines = f.readlines()
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply utils/check_content_style.py:45]: By sorting the files I was just trying to make it easier for content creators to fix issues, since they probably see the files in alphabetical order.

File: utils/check_content_style.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+def print_error(file, line, reason):
+	error_list.append((line, "Formatting error in " + file + " line " + str(line) + ": " + reason))
+	global errors
+	errors += 1
+
+
+if __name__ == '__main__':
+	check_content_style()
+	if errors > 0:
+		text = "Found " + errors.__str__() + " formatting " + ("error" if errors == 1 else "errors")
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply utils/check_content_style.py:168]: Yes, that is already fixed in my local fork, I just didn't push it to github yet :)

File: utils/check_content_style.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	# this program. If not, see <https://www.gnu.org/licenses/>.
+	\
+	""".splitlines()
+	text_index = 0
+	for line in lines:
+		if re.search(first, line):
+			text_index += 1
+		else:
+			break
+	if text_index == 0:
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply utils/check_content_style.py:98]: Yeah, this is one of the worst-looking parts of the code. The thing is, the `first` regex may match multiple times, since copyright headers can contain more than one copyright holders, so I can't assume it is only going to be the first item in the list.

File: utils/check_content_style.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+			previous_indent = indent
+
+
+# Counts the number of tabs this line of text is indented with.
+# Parameters:
+# text: the text to count indents in
+def count_indent(text):
+	count = 0
+	for char in text:
+		if char == '\t':
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply utils/check_content_style.py:155]: I accept if you say that is valid, however that feature doesn't actually seem to be used anywhere, making tabs a de facto standard.

File: utils/check_content_style.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+regexes = {
+	# Matches any space that is preceeded only by tabs or " or ` characters,
+	# except if it immediately follows a " or ` character.
+	"^[\\t`\"]*(?<![`\"]) ": "indentations should use tabs only",
+	# Matches multiple space characters.
+	"  ": "multiple consecutive space characters",
+	"\\s$": "trailing whitespace character"
+}
+# Precompile regexes
+regexes = {re.compile(regex): description for regex, description in regexes.items()}
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply utils/check_content_style.py:40]: If you feel like that is more readable, sure.

File: utils/check_content_style.py (hunk 3 above)
Comment (tibetiroka) [reply utils/check_content_style.py:51]: Oh that's an interesting take. I removed line endings because they messed with the regexes, but I suppose we could include additional tests for checking the line separators.

File: utils/check_content_style.py (hunk 4 above)
Comment (tibetiroka) [reply utils/check_content_style.py:56]: Okay, I think I can combine that with your previous 'no globals' comment to make something functional

File: utils/check_content_style.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tehhowch) [reply utils/check_content_style.py:40]: Even more readable would be separation of concerns, where one module is responsible for setting up and exporting the expressions and another is responsible for using them 🙂

File: utils/check_content_style.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tehhowch) [reply utils/check_content_style.py:155]: The very next place this utility would be used is by plugins, who can and should be allowed to use whatever indentation style they want. Therefore baking in "tabs are the only acceptable way to indent" is not ok.

**Comment by tibetiroka** [comment]
I don't have a lot of time today, but later on I'll go through the rest of those suggestions and fix them both here and in the code checked script as well.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -515,3 +515,20 @@ jobs:
...
+
+  check-content-style:
+    needs: changed
+    if: needs.changed.outputs.data == 'true'
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
+    - name: set up python environment
+      uses: actions/setup-python@v2
```
Comment (tehhowch) [inline .github/workflows/ci.yml:528]:
```suggestion
      uses: actions/setup-python@v4
```

File: utils/check_content_style.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+	"^[\\t`\"]*(?<![`\"]) ": "indentations should use tabs only",
+	# Matches multiple space characters, unless they are followed by a '#', or if the text is a tooltip.
+	"(?<!^tip .*)(  $|  [^ #])": "multiple consecutive space characters",
+	# Matches any whitespace character at the end of the string.
+	"[^\\s]\\s+$": "trailing whitespace character",
+	# Matches any backticks at the end of the line that is
+	# preceded by the ending of a sentence ('?', '!', '.', ';', ':', ')', '"', ''', '>') and a whitespace.
+	# The same check fails if used on quotation marks, due to the irregular nature of 'word'.
+	"[?!.;:\\)\"'>] `$": "unnecessary space before closing backtick"
+}.items()}
```
Comment (tehhowch) [inline utils/check_content_style.py:34]: You can forgo the construction of an inner dict by iterating a list of 2-element lists

File: utils/check_content_style.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tehhowch) [reply utils/check_content_style.py:45]: You can group the errors returned by the call to `check_content_style(lines)` in the calling context, and there you can label the output with the filename. That will be even more readable (and more performant, since you won't be interleaving IO writes with processing and can invoke this method on many files at once).

File: utils/check_content_style.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tehhowch) [reply utils/check_content_style.py:165]: The game itself will warn about mixed indentation in a file, so you need not worry about that case here.

File: utils/check_content_style.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+if __name__ == '__main__':
+	error_count = check_content_style()
+	if error_count > 0:
+		text = "Found " + str(error_count) + " formatting " + ("issue" if error_count == 1 else "issues") + "."
+		print(text)
+		exit(1)
+	else:
+		print("No formatting issues found.")
+	exit(0)
```
Comment (tehhowch) [inline utils/check_content_style.py:185]: The only place any file reads or console writes should occur is in this `'__main__'` block. All other methods should be taking a list of lines and returning a list of errors.

File: utils/check_content_style.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tibetiroka) [reply utils/check_content_style.py:165]: Okay, I guess I'll just remove that check then.

**Comment by tibetiroka** [comment]
@tehhowch I've been thinking about what you said about plugins using style checker. I don't think this style checker would ever reach a state where it can be reasonably used for plugins.

The indentation checks are going to be removed, the double space checks need some serious work (although it will be functional), the trailing whitespace and empty line checks are based standards we don't even enforce in our own content and copyright checks are useless for plugins, leaving us with the single ascii check. We couldmaybe create a small one-liner script for that if necessary, but nothing like the script in this pr.

Basically, my question is this: do you think plugins would use this script, if it was part of ES? If so, it's probably best to just close this pr and leave this as a manual tool for content reviewers, if they choose to use it.
